src/main/java/org/jenkinsci/maven/plugins/hpi/JenkinsHpiDependencyCollector.java

```diff
@@ -0,0 +1,375 @@
+package org.jenkinsci.maven.plugins.hpi;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.apache.maven.RepositoryUtils;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.repository.ArtifactRepository;
+import org.apache.maven.artifact.resolver.filter.ArtifactFilter;
+import org.apache.maven.model.Dependency;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectBuildingRequest;
+import org.apache.maven.shared.dependency.graph.DependencyCollectorBuilderException;
+import org.apache.maven.shared.dependency.graph.DependencyNode;
+import org.apache.maven.shared.dependency.graph.internal.ConflictData;
+import org.apache.maven.shared.dependency.graph.internal.DefaultDependencyNode;
+import org.apache.maven.shared.dependency.graph.internal.DirectScopeDependencySelector;
+import org.apache.maven.shared.dependency.graph.internal.VerboseJavaScopeSelector;
+import org.eclipse.aether.DefaultRepositorySystemSession;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.artifact.ArtifactTypeRegistry;
+import org.eclipse.aether.collection.CollectRequest;
+import org.eclipse.aether.collection.CollectResult;
+import org.eclipse.aether.collection.DependencyCollectionException;
+import org.eclipse.aether.collection.DependencyGraphTransformer;
+import org.eclipse.aether.collection.DependencySelector;
+import org.eclipse.aether.graph.DependencyVisitor;
+import org.eclipse.aether.graph.Exclusion;
+import org.eclipse.aether.util.artifact.JavaScopes;
+import org.eclipse.aether.util.graph.manager.DependencyManagerUtils;
+import org.eclipse.aether.util.graph.selector.AndDependencySelector;
+import org.eclipse.aether.util.graph.transformer.ConflictResolver;
+import org.eclipse.aether.util.graph.transformer.JavaScopeDeriver;
+import org.eclipse.aether.util.graph.transformer.NearestVersionSelector;
+import org.eclipse.aether.util.graph.transformer.SimpleOptionalitySelector;
+import org.eclipse.aether.util.graph.visitor.TreeDependencyVisitor;
+import org.eclipse.aether.version.VersionConstraint;
+
+/**
+ * <p>
+ *     Project dependency raw dependency collector with Jenkins hpi filter need
+ * </p>
+ * <p>adaptation of some classes from
+ * https://github.com/apache/maven-dependency-tree/tree/master/src/main/java/org/apache/maven/shared/dependency/graph/internal
+ * </p>
+ */
+@Named
+public class JenkinsHpiDependencyCollector {
```

Comment:
the only change is the removal of this `new OptionalDependencySelector()` this is our problem  https://github.com/apache/maven-dependency-tree/blob/fa81a70f7c6e3e37b77f57896156051273c0b2f3/src/main/java/org/apache/maven/shared/dependency/graph/internal/DefaultDependencyCollectorBuilder.java#L108 

but to keep using the mainstream code and not having to maintain a sort of forked class I have made this PR https://github.com/apache/maven-dependency-tree/pull/28
this will help us